AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -447,10 +447,9 @@ - (void)playMusic:(BOOL)continuous {
             // rendring and capturing.
             [self initializeCapturer];
             [self initializeRenderer];
-            
-            TVIAudioDeviceContext *context = NULL;
-            [self startRendering:context];
-            [self startCapturing:context];
+        
+            [self startRendering:self.renderingContext->deviceContext];
+            [self startCapturing:self.capturingContext->deviceContext];
```

Comment:
This handles the case of playing music while the device is already playing or recording. 👍 

But, are there other transitions that could still be invalid and crash?

The case I am most concerned about is whenever `stopRendering` or `stopCapturing` is called and the `TVIAudioDeviceContext` becomes invalid. When we are really stopping the AVAudioUnit (final caller) there is a synchronization point after which there is a guarantee of no more callbacks. But the first invocation of `stopRendering` or `stopCapturing` might be unsafe since it does not result in a synchronization with the core audio thread.

Reply:
@ceaglest in our example we do never invoke `stopCapturing` or `stopRendering` explicitly. So we don't have a chance of device context becoming NULL while AudioUnit Callbacks are invoked. I tested it multiple times and did not see any crash.

Reply:
For some context it's not typical for the application to call `stopCapturing` or `stopRendering`. These methods are invoked when either the the last local audio is unpublished, the last remote audio is unsubscribed or the last PeerConnection is closed.